worker/caasunitprovisioner/application_worker.go

```diff
@@ -84,17 +122,31 @@ func (aw *applicationWorker) loop() error {
 				if err != nil {
 					return errors.Trace(err)
 				}
-				if _, ok := unitWorkers[unitId]; ok || unitLife == life.Dead {
-					// Already watching the unit. or we're
-					// not yet watching it and it's dead.
-					continue
+
+				if unitLife == life.Dead {
+					aliveUnits.Remove(unitId)
+				} else {
+					aliveUnits.Add(unitId)
 				}
-				w, err := newUnitWorker(unitId, aw.broker, aw.containerSpecGetter)
-				if err != nil {
-					return errors.Trace(err)
+				if !aw.brokerManagedUnits {
+					// For Juju managed units, we start a worker to manage the unit.
+					if _, ok := unitWorkers[unitId]; ok || unitLife == life.Dead {
+						// Already watching the unit. or we're
+						// not yet watching it and it's dead.
+						continue
+					}
+					w, err := newUnitWorker(aw.application, unitId, aw.containerBroker, aw.containerSpecGetter)
+					if err != nil {
+						return errors.Trace(err)
+					}
+					unitWorkers[unitId] = w
+					aw.catacomb.Add(w)
 				}
-				unitWorkers[unitId] = w
-				aw.catacomb.Add(w)
+			}
+			// Now we have processed all the changed units, notify the
+			// worker of those that are still alive.
+			if aw.brokerManagedUnits {
+				aw.aliveUnitsChan <- aliveUnits.Values()
```

Comment:
No naked channel sends. What happens during agent termination? And also, this prevents processing further unit changes until the channel is ready to receive.

I think you want to do something like this:

```go
var aliveUnitsChan chan []string
...
select {
...
case units, ok := <-uw.Changes():
    ...
    if aw.brokerManagedUnits {
        aliveUnitsChan = aw.aliveUnitsChan
    }
case aliveUnitsChan <- aliveUnits.Values():
    aliveUnitsChan = nil
}
```

Reply:
fixed